[HarfBuzzSharp] font funcs

Gillibald opened the pull request.

mattleibow commented:
@Gillibald looking good. I made a PR that will reduce the code to 2 lines per function: https://github.com/mono/SkiaSharp/pull/882

Feel free to merge that into your PR and then I can close it. If it works, we can merge with your PR. If it doesn't then you can adjust it on your PR to fit whatever is needed.

Gillibald commented:
Will have a look shortly 👍

Gillibald commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,553 @@
+				var unicodes = new ReadOnlySpan<uint> ((void*)firstUnicode, (int)unicodeStride);
+				var glyphCount = del.Invoke (null, null, count, unicodes, out var glyphs);
+
+				fixed (uint* glyphsPtr = glyphs) {
```
Not sure how to do this properly. If the delegate creates an array on native side everything is fine but if the array is managed this will break.

Gillibald commented:
@mattleibow Your PR helped a lot. This now covers all font funcs. I will write some unit tests tomorrow.

Gillibald commented:
Looks like something isn't right with the interop code or the way I wrote that failing unit test.

Gillibald commented on binding/HarfBuzzSharp.Shared/Font.cs:
```diff
@@ -17,11 +18,34 @@ public Font (Face face)
 
+		public Font Parent {
+			get {
+				var parent = HarfBuzzApi.hb_font_get_parent (Handle);
```
Introduce a backing field so we can store the .Net instance when a sub font is created by the binding

Gillibald commented on binding/HarfBuzzSharp.Shared/HarfBuzzApi.cs:
```diff
@@ -118,6 +166,15 @@ internal class HarfBuzzApi
+		[DllImport (HARFBUZZ, CallingConvention = CallingConvention.Cdecl)]
+		public extern static hb_bool_t hb_font_get_glyph_name (hb_font_t font, hb_codepoint_t glyph, out StringBuilder name, out uint size);
+		[DllImport (HARFBUZZ, CallingConvention = CallingConvention.Cdecl)]
+		public extern static hb_bool_t hb_font_get_glyph_from_name (hb_font_t font, string name, int len, out hb_codepoint_t glyph);
```
UnmanagedType.LPStr attribute missing

Gillibald commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,553 @@
+		uint pointIndex, out int x, out int y, IntPtr context);
+	internal delegate bool GlyphNameProxyDelegate (IntPtr font, IntPtr fontData, uint glyph, out StringBuilder name,
+		out uint size, IntPtr context);
+	internal delegate bool GlyphFromNameProxyDelegate (IntPtr font, IntPtr fontData, StringBuilder name,
```
change name to string with UnmanagedType.LPStr

Gillibald commented on binding/HarfBuzzSharp.Shared/HarfBuzzApi.cs:
```diff
@@ -118,6 +166,15 @@ internal class HarfBuzzApi
+		[DllImport (HARFBUZZ, CallingConvention = CallingConvention.Cdecl)]
+		public extern static hb_font_t hb_font_get_parent (hb_font_t font);
+		[DllImport (HARFBUZZ, CallingConvention = CallingConvention.Cdecl)]
+		public extern static hb_bool_t hb_font_get_glyph_name (hb_font_t font, hb_codepoint_t glyph, out StringBuilder name, out uint size);
```
Change StringBuilder to string with UnmanagedType.LPStr

mattleibow replied:
I don't think the size is an `out`. It looks to be calculated/assumed somewhere first: https://github.com/harfbuzz/harfbuzz/blob/2.5.2/src/hb-font.h#L507-L510

Gillibald commented on binding/HarfBuzzSharp.Shared/Font.cs:
```diff
@@ -211,6 +253,14 @@ public GlyphExtents GetGlyphExtents (uint glyph)
+				: glyph.ToString ();
+
+		public uint GetGlyphFromName (string name) =>
+			HarfBuzzApi.hb_font_get_glyph_from_name (Handle, name, name.Length, out var glyph) ? glyph : 0;
```
`len = - 1`

Gillibald commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,553 @@
+		out GlyphExtents extents, IntPtr context);
+	internal delegate bool GlyphContourPointProxyDelegate (IntPtr font, IntPtr fontData, uint glyph,
+		uint pointIndex, out int x, out int y, IntPtr context);
+	internal delegate bool GlyphNameProxyDelegate (IntPtr font, IntPtr fontData, uint glyph, out StringBuilder name,
```
No out parameters

Gillibald commented:
@mattleibow This PR changes a lot more than just adding font funcs to the binding. I also added missing members and removed some overloads that take a `signed integer` as glyph id. HarfBuzz itself and all the binding methods return `an unsigned integer` as `glyph id` so it makes sense to remove these overloads. Skia and most other platforms use `ushort` as glyph id.

I also changed some methods to follow the Try-pattern because HarfBuzz itself works that way and it doesn't make sense to get a value from HarfBuzz all the time. This lets the consumer know that something wasn't found etc.

Gillibald commented:
@mattleibow Sry for the refactoring of some parts. I thought better change it now before the API | ABI is frozen. Hopefully, this doesn't take too long to review.

mattleibow commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,536 @@
+		private static bool FontExtentsProxyImplementation (IntPtr font, IntPtr fontData, out FontExtents extents, IntPtr context)
+		{
+			var multi = DelegateProxies.Get<GetMultiDelegateDelegate> (context, out _);
+			var del = (FontExtentsDelegate)multi.Invoke (typeof (FontExtentsDelegate));
```
Just pushed a commit to master that will compress these lines:

```csharp
 var del = GetMulti<FontExtentsDelegate> (context, out _);  
```

Gillibald replied:
Will make the needed changes 👍

Gillibald commented:
@mattleibow I have implemented the delegate for font_data(user_data) it seems to work. Feel free to rename things etc.

mattleibow commented:
@Gillibald this is looking really good.

mattleibow commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,530 @@
+		out GlyphExtents extents, IntPtr context);
+	internal delegate bool GlyphContourPointProxyDelegate (IntPtr font, IntPtr fontData, uint glyph,
+		uint pointIndex, out int x, out int y, IntPtr context);
+	internal unsafe delegate bool GlyphNameProxyDelegate (IntPtr font, IntPtr fontData, uint glyph, char* name,
```
Is there a reason why this is `char*` instead of `Span<T>`? Or even a `string`?

Gillibald replied:
I will change it to string. Wasn't sure at the beginning what HarfBuzz is expecting here.

mattleibow replied:
I just pushed the changes - let me make some comments

mattleibow commented:
Something is failing on Windows. I ran the Windows test 9 times on CI and they all failed. I also fails locally. Not sure what it is... Will keep looking.

mattleibow commented:
I think I may have discovered it. I build harfbuzz in Debug mode and got a nice exception:

> Run-Time Check Failure #0 - The value of ESP was not properly saved across a function call.  This is usually a result of calling a function declared with one calling convention with a function pointer declared with a different calling convention.

Just needed to add an attribute to the proxy delegates:

```csharp
[UnmanagedFunctionPointer (CallingConvention.Cdecl)]
```

mattleibow commented on binding/HarfBuzzSharp.Shared/DelegateProxies.font.cs:
```diff
@@ -0,0 +1,223 @@
+
+	public delegate bool GlyphNameDelegate (Font font, object fontData, uint glyph, out string name);
+
+	public delegate bool GlyphFromNameDelegate (Font font, object fontData, string name, out uint glyph);
```
Both `GlyphNameDelegate` and `GlyphFromNameDelegate` now use strings.

mattleibow commented on binding/HarfBuzzSharp.Shared/DelegateProxies.font.cs:
```diff
@@ -0,0 +1,223 @@
+
+			var nameSpan = realName.AsSpan ();
+			var bufferSpan = new Span<char> (nameBuffer, size);
+			nameSpan.CopyTo (bufferSpan);
```
Copy the managed string to the native buffer.

mattleibow commented on binding/HarfBuzzSharp.Shared/DelegateProxies.font.cs:
```diff
@@ -0,0 +1,223 @@
+
+			var actualName = len < 0
+				? new string (name)
+				: new string (name, 0, len);
```
Create a string from the buffer, keeping in mind that a `-1` length is an null-terminated string.

mattleibow commented on binding/HarfBuzzSharp.Shared/Font.cs:
```diff
@@ -163,54 +124,99 @@ public unsafe int[] GetVerticalGlyphAdvances (int count, IntPtr firstGlyph)
+						name = string.Empty;
+						return false;
+					}
+					name = new string (first);
```
Instead of checking for a null character, let the string constructor do the work.

mattleibow commented on binding/HarfBuzzSharp.Shared/Font.cs:
```diff
@@ -163,54 +124,99 @@ public unsafe int[] GetVerticalGlyphAdvances (int count, IntPtr firstGlyph)
+			try {
+				fixed (char* first = buffer) {
+					HarfBuzzApi.hb_font_glyph_to_string (Handle, glyph, first, buffer.Length);
+					return new string (first);
```
Avoid the manual loop, use the constructor.

mattleibow commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,270 @@
+			if (IsImmutable)
+				throw new NotSupportedException ($"{nameof (FontFunctions)} is immutable and can't be changed.");
+			if (del == null)
+				throw new ArgumentException (nameof (del));
```
These actually need to be null arg exceptions...

mattleibow commented on binding/HarfBuzzSharp.Shared/FontFunctions.cs:
```diff
@@ -0,0 +1,270 @@
+		public void SetVerticalGlyphAdvanceDelegate (GlyphAdvanceDelegate del, ReleaseDelegate destroy)
+		{
+			if (IsImmutable)
+				throw new NotSupportedException ($"{nameof (FontFunctions)} is immutable and can't be changed.");
```
This should be InvalidOperationException.

Gillibald commented:
@mattleibow Thanks for cleaning this up. Looks so much better now. I can learn a lot from you. Just tried to get this in fast. Helps a lot to have a look from another perspective.

Gillibald commented on binding/HarfBuzzSharp.Shared/DelegateProxies.cs:
```diff
@@ -79,7 +76,7 @@ private static void ReleaseDelegateProxyImplementation (IntPtr context)
 		{
 			var del = GetMulti<GetTableDelegate> (context, out var gch);
-			var blob = del.Invoke (null, tag, null);
+			var blob = del.Invoke (null, tag);
```
Do we want to use the UserDataDelegate here as well? If not we should remove the face parameter.

mattleibow replied:
Good catch. Let me fix that.

Gillibald commented:
https://github.com/harfbuzz/harfbuzz/releases/tag/2.5.3

I guess we have to retarget again

mattleibow commented:
Thanks for doing all the work for the functions. I probably wouldn't have been able to get them in as fast as you did - and I probably would have gotten it wrong. Nice to bounce all the ideas off you so that it is a better API.

mattleibow commented:
> I guess we have to retarget again

lol It has only been a few days and they are already pushing out releases :)

Gillibald commented:
Luckily it is trivial to update the submodule

mattleibow commented:
I had a look at the diff between 2.5.2 and 2.5.3 and there are no new files, so that is nice.

mattleibow commented:
@Gillibald let me know if these changes still work for the way you consume the APIs. We can merge if it is all good. I don't think I changed anything on the public API front, except the strings in the delegates.

Gillibald commented:
Haven't found the time yet to adopt these changes in my code. Should be fine but we can wait until I have made these changes if you want.

Gillibald commented:
Okay just tested against this PR and everything looks good 👍
